Integration with auto formatters

emillon opened the pull request:
This adds a "fmt" extension in dune-project files. When used, it will setup a `@fmt` alias that will call `ocamlformat` on ocaml source code, and `refmt` on reason source code. The tools are not configured by dune.

### :memo: TODO

- [x] doc
- [x] squash
- [x] benchmark
- [x] fix the plumbing when parsing
- [x] generate a .formatted directory, computing the rules lazily
- [x] add dependencies to ocamlformat files
- [x] ~~pass `--name`~~ (superseded by dependencies)

hhugo commented on src/ocamlformat.ml:
```diff
@@ -0,0 +1,78 @@
+    let formatted = formatted_path path in
+    let format_action =
+      let args = [sv flag; dep path] in
+      Action.Unexpanded.Redirect
```
Is this better than having ocamlformat write down the file directly ?

hhugo commented on src/dune_file.ml:
```diff
@@ -678,6 +678,7 @@ module Buildable = struct
     ; ocamlopt_flags           : Ordered_set_lang.Unexpanded.t
     ; js_of_ocaml              : Js_of_ocaml.t
     ; allow_overlapping_dependencies : bool
+    ; ocamlformat : bool
```
Should this be a property of libs and binary or just of the current directory ?
also, You might want to be able to add exceptions, useful for automatically generated source files

emillon replied:
it's now possible to use OSL to perform exclusions.
per directory would work, but we already have a mapping between files and modules in buildables, so that seems to be the correct abstraction (at least it's the simplest one to start)

hhugo commented on src/lib_rules.ml:
```diff
@@ -374,6 +374,7 @@ module Gen (P : Install_rules.Params) = struct
         ~lib_name:(Some lib.name)
         ~dir_kind
     in
+    Ocamlformat.gen_rules sctx lib.buildable ~dir ~scope modules;
```
Is this used for ocaml files only or also for reasonML files ?

emillon replied:
now, this ignores reason files

hhugo commented:
Well, you have some early review now

emillon commented:
Well, nothing a good night of sleep couldn't fix. I just passed the basename and it now works. I'll ping the PR when it's ready for more feedback :) Sorry for the noise! (& thanks for the early review!)

emillon commented:
I believe that this is ready for a round of review.

Some outstanding problems/questions:
- is it ok to just pick a new alias? `@format` is pretty specific, but this would break user code if some dune files already use this alias.
- this is enabled within buildable stanzas (library/executable), but per directory (ie per dune file) would work as well. It's just harder to list the source files per directory compared to per stanza, I believe.
- in current versions of ocamlformat (0.7), there is a risk of configuration proliferation: if there is no `.ocamlformat` file in the source tree, a more global one will be used (see ocaml-ppx/ocamlformat#393). Should we wait until a release is cut and hardcode `--root`? Or detect the version with best effort and add the flag if supported?

Thanks!

samoht commented:
2 remarks:
- I think I prefer `dune build @fmt` or `dune fmt` instead of `@format` (e.g. like `go fmt`)
- Could this just be enabled by default on all the libraries/binaries (e.g. without the need for a specific stanza) so you could just write `dune fmt` on an existing project and it will just reformat your project?

hcarty commented:
Once this is in, would it be possible to have the same support for Reason syntax and `refmt`? It would be very helpful to be consistent between the two syntaxes.

Aa far as enabling formatting by default, ocamlformat can be disabled per-file with an attribute. So that provides an option for users who need a file or few to remain untouched by `dune fmt`.

emillon commented:
Thanks for your feedback!

I changed this so that it works as a stanza, per directory. This uses `Dir_contents`, so as I understand it, will only work for in-tree source files (which seems to be the correct thing to do). I kept the OSL field for exclusions.

I suggest the following plan to generalize this a bit:
- rename the stanza `(format)`
- call `refmt` on reason files
- auto-format dune files (using what's called `dune unstable-fmt` at the moment, renaming it to `dune format-dune-file`?)
- add some configuration option to determine what autoformatters are enabled, so that one can e.g. only format dune files. This might replace the current OSL field.
- add a new command `dune fmt` which calls `dune build @format` - reusing the current convention for `dune runtest`. Should we call the alias `@fmt`, then? (the `command fmt` convention is well adopted now)

This plan would obviously span several PRs, but it's great if we all agree on the direction to take.

avsm commented:
Why do we need a `(format)` stanza at all?  It seems useful to just make this alias available unconditionally on valid OCaml files.  One cool usecase would be to then run `dune fmt` on a large monorepo like the OCaml Platform one and have it format all the different projects consistently :)

diml commented:
For the CLI, I suggest to have a single sub-command to format all kinds of files. `dune fmt` sounds good. We can make it take an optional filename to reformat a single file. I agree that the alias and command name should match, so `dune fmt` and `@fmt` sounds good.

I agree with @avsm that we shouldn't need one stanza per directory, except when there are exceptions. We might want to allow to opt-in or opt-out via the dune-project file though, so that in a mono-repo dune can skip projects that don't want to be auto-formatted. I suggest to make it opt-in via the dune-project file for now, to be conservative, and opt-out with `(lang dune 2.0)`.

emillon commented:
OK! Let's make this less hidden and more global. I'll deal with the CLI part in a separate PR.

I am not sure how to add new features through `dune-project`. At the moment there is only `lang`, `name` and `version` in these files. Adding `fmt` here seems very specific. Should I add a new syntax and `(using fmt)`? Does `(using)` enable new syntax in dune-project itself?

Thanks!

diml commented:
Yh, the idea is to add `(using fmt <version>)`. Currently `(using)` only enables new stanzas in `dune` files, though you can pass parameters to the `using` stanza:

```
(using fmt 1.0
 (enabled_for ocaml reason dune))
```

In this case, rather than or in addition to adding a new stanza in `dune` files, we probably want to also be able to just check for the presence of this stanza in the dune-project file. I suppose we could change the API of `Dune_project.Extension` as follow:

```ocaml
module Extension : sig
  type 'a t

  (** [register id parser] registers a new extension. Users will
      enable this extension by writing:

      {[ (using <name> <version> <args>) ]}

      in their [dune-project] file. [parser] is used to describe
      what [<args>] might be. *)
  val register
    :  ?experimental:bool
    -> Syntax.t
    -> ('a * Stanza.Parser.t list) Dsexp.Of_sexp.t
    -> 'a t
end

val get_extension : t -> 'a Extension.t -> 'a option
```

emillon commented:
I just did that. For now it just returns the loc but I'll extend that to refmt and add `enabled_for`.

diml commented on src/format_rules.ml:
```diff
@@ -0,0 +1,82 @@
+    let msg = "Cannot find ocamlformat, skipping.\n" in
+    add_alias (Echo [sv msg])
+  | Some ocamlformat ->
+    let files = Dir_contents.text_files dir_contents in
```
This includes generated files. We really need to look into the file tree for the actual files to reformat.

emillon replied:
Done. I used `Path.drop_build_context_exn` which works, is that the correct function to use?

diml replied:
yep

diml commented on src/dune_project.ml:
```diff
@@ -215,6 +215,7 @@ module Extension = struct
     }
 
   let extensions = Hashtbl.create 32
+  let instanciated_extensions = Hashtbl.create 32
```
This should be per-project, otherwise it won't work when there are several projects in the same workspace.

emillon replied:
It's now stored in the project.

diml commented:
BTW, could you benchmark this change? The platform repository should be a good test case. Given that this more work that we will do systematically, we need to know what the cost is.

emillon commented:
Yes, I'm adding a cache so that we only call which on each formatter as well. But a benchmark will be useful!

emillon commented:
I:
- added a lazy around the `which` calls
- extended to `refmt`

Next I'll add the configuration syntax and make the instanciation table per project.

emillon commented:
About testing: rather than installing ocamlformat & refmt, I was thinking that I could bring "fake" binaries in PATH (e.g. prepending each line with a prefix + displaying sys.argv in a header), what do you think about that?

emillon commented:
I'm not sure how to implement passing `('a * Stanza.Parser.t list) Dsexp.Of_sexp.t` to `Extension.register`. The problem is that extensions are stored into a hash table, which has to stay homogeneous. What is the best way to do it? store the ast to parse it later?

emillon commented:
I think I did it! I used `Univ_map` to store the arguments. It seems to be more complicated than it should, so I'm pretty curious about what could be improved. Thanks!

diml commented:
Regarding the cost of this feature in terms of performance, what I have in mind is more the cost of adding one more rule + one more alias dependency per .ml file. Currently Dune generates all the rules for a given directory at once, so adding one rule per .ml is costly. Eventually, we'll allow to define more fine grained notion of partitions, such as directory+glob, but currently this is not supported.

To measure the cost, you can do the following: take a big repository, such as the OCaml platform one, build it and then measure the cost of a null build. That will give you the cost of computing all the rules.

If the cost is too high, then instead of generating `foo.ml.corrected` files, we should generate `.formatted/foo.ml` files. I.e. move the rules to a sub-directory so that they are only computed if needed. BTW, it might make sense to do that anyway, so that `.formatted` files are not included in `(glob_files *)`.

rgrinberg commented on src/dune_file.ml:
```diff
@@ -650,6 +650,62 @@ end
+
+  let to_sexp {loc; enabled_for} =
+    Sexp.List
+      [ List [Atom "loc"; Loc.to_sexp loc]
```
It's a bit weird to include the location in the sexp I think. This is just for debugging, right?

emillon replied:
It's required because this type goes into a `Univ_map.t` I think it's OK to remove it.

rgrinberg commented on src/dune_file.ml:
```diff
@@ -650,6 +650,62 @@ end
+    let%map loc = loc
+    and enabled_for = record enabled_for_field
+    in
+    ({loc; enabled_for}, [])
```
The signature of `dparse` should usually be `t Dsexp.Of_sexp.t`

jberdine commented:
I'm not sure I have fully understood the latest state of this PR, but would it make sense to discuss a bit to try to avoid a situation where people need to duplicate config between dune and ocamlformat?

In order to avoid duplicating logic in multiple editor integrations, ocamlformat is moving in the direction of (and very close to) being able to be invoked on any source file and only process it if configured to do so. That is, we don't want to have the editor integration deciding whether or not ocamlformat should be enabled, but to just always invoke it and let it do nothing when appropriate.

The current mechanisms for configuring when it is enabled are a per-directory file and attributes within individual files. It could very well be that the dune configuration for when to enable ocamlformat is better, but it would be better if there is just one config. I don't know if it is feasible for the editor integrations for ocamlformat to rely on asking dune to invoke ocamlformat, due to situations like wanting to format unsaved files, etc. I also don't know if it would be feasible for dune to directly consume the ocamlformat configuration for where it is enabled, perhaps if it was stored in a different form. Or perhaps it would make sense for ocamlformat to ask dune for the dune config and read it. Or for ocamlformat to look in `_build/.jbuilds/<context>/<dir>/dune` itself. (Aside, what's up with that dir still being named `.jbuilds`?)

Thoughts?  Cc @hhugo @gpetiot

emillon commented:
@jberdine at the moment, what this PR roughly does is `for i in **/*.ml **/*.mli ; do ocamlformat $i -o $i.formatted; done`. This is not directly intended to be used by editors, but rather by CI checks and as convenience for contributors who don't have editor integration set up, for example. This is a pretty generic mechanism that is not specific to ocamlformat - this PR includes refmt support as well, and later will be extended to dune files. It is possible that this gets extended to C files one day.

As for the configuration, dune does not look for `.ocamlformat` files. If a file is excluded by an ocamlformat config file, the difference will just be a no-op. On the other hand, if ocamlformat provides a way to return a list of files that need to be formatted, we can use that so that we don't needlessly iterate on ignored files. But I see that more as an optimization.

For editor integration, I think it's a different issue completely. IMO, this is not dune's job to sit between the editor and `ocamlformat`, and the detection feature you've implemented (recurse up until `.git` or `dune-project`) seems the way to go.

This separation of concerns seems pretty fine to me - in particular, there's no duplication of configuration, and no duplication of work between dune and ocamlformat. What do you think of this plan?

jberdine commented:
@emillon that sounds good, thanks! It could be that there is some worthwhile optimization to avoid needlessly not-reformatting a bunch of files (e.g. with a large amount of vendored code) but it seems premature to worry about that now. I guess that this is respecting e.g. `ignored_subdirs`, etc.

diml commented:
Currently this feature is a per-project opt-in, so vendored code shouldn't be an issue. One enables it by writing `(using fmt 1.0)` in the `dune-project` file.

BTW, this discussion makes me think that the rules in this PR should declare a dependency on the `.ocamlformat` file when one is found in the source tree, otherwise it won't be copied to `_build/default`.

hhugo commented:
I don't think you need `.ocamlformat` files to be copied to `_build`. The ocamlformat invocation should probably mention the original file name (with `--name`) and that should be enough to have ocamlformat look at the right `.ocamlformat` files.

hhugo commented:
Should ocp-indent be supported in the same ways ?

jberdine commented:
^^ but perhaps `.ocamlformat` should be a dependency anyway, in order to reformat when the config changes.

emillon commented:
I don't think ocp-indent should be supported, at least in a first version. This keeps the configuration simple and gives a good incentive to use `ocamlformat` for new projects.

emillon commented on src/gen_rules.ml:
```diff
@@ -211,12 +211,20 @@ module Gen(P : Install_rules.Params) = struct
 
   let gen_rules dir_contents
         { SC.Dir_with_jbuild. src_dir; ctx_dir; stanzas; scope; kind } =
+    gen_format_rules sctx ~dir:ctx_dir ~scope;
```
What would be a good place to call this setup code? It's a bit difficult to understand what the different parts are doing.

diml replied:
In the last `gen_rules` function in this module, in the `match Dir_contents.kind dir_contents with`, we iterate over all directories of the group at this point.

The code is a bit complex indeed, it could use some refactoring

emillon replied:
I tried to call it just before that match. Is this the way to go? Also there's no `scope`, so I need to rebuild it from the `sctx` here.

diml replied:
We also need to call it for every sub-directory in the case `Group_root`.  Looking at this again, we could also call it in the `gen_rules` function that comes just before since this one is called once for every directory in the group.

diml commented:
>
> Just did that. I didn't manage to use a workspace-global binary though:
> the system-wide one is used when the public executable is defined outside
> of the project.
>

That doesn't seem right, we'll need to look into that.

emillon commented:
That's fixed, I was not setting up the workspace correctly (I was passing `--root`). It works as expected.

emillon commented on src/dune_project.mli:
```diff
@@ -56,6 +56,8 @@ module Lang : sig
 end
 
 module Extension : sig
+  type 'a key
```
There's already a `t` in this module, which is a different type. I'll rename it, it will make more sense from the outside.

hhugo commented:
Ocamlformat is still very young and IMHO still lacking some tooling to be usable without friction. In particular, I'ld like to be able to rebase commits across style changes without having to resolve artificial conflicts. It seems odd to me not support solid existing tools like ocp-indent. I don't think ocp-indent has been super-seeded by ocamlformat yet.

diml commented on src/format_rules.ml:
```diff
@@ -0,0 +1,82 @@
+
+    let ocaml kind =
+      if config_includes config Ocaml then
+        let exe = resolve_program "ocamlformat" in
```
We could cache this outside of the `Path.Set.iter`, to avoid recomputing it for every file.

emillon commented:
Regarding ocp-indent, only allowing one formatter per language keeps this feature simple. If several formatters are possible, this prompts the questions of how to configure ocp-indent for certain files and ocamlformat for other ones; how to use both on the same file; define integrations with custom formatting tools, etc. It is too soon for this.

I agree that ocamlformat is less mature than ocp-indent, but eventually it will replace it. I don't think we're far from there. In particular, it's a good choice to make for a new project.

And as usual in open source, "no" is temporary but "yes" is forever. It's easier to start with a precise scope and expand it later, rather than having to support a very flexible system that's built on assumptions that don't hold a few releases later. I encourage you to open an issue where we can discuss this for later improvements. Thanks!

hhugo commented:
> I don't think we're far from there. In particular, it's a good choice to make for a new project.

These two claims are really not obvious to me.

I agree that having a single tool for formatting OCaml source seems better. At JaneStreet, that tool is neither ocp-indent nor ocamlformat. It's a separate tool responsible to managing other formatters. (cc @gpetiot and @samoht)

diml commented:
From the point of view of Dune, it definitely makes our lives easier if there is a single auto-formatter to consider per language, and if we can delegate all the questions about how it should be done to this tool.

diml commented:
So I tend to agree with @emillon that we should only support `ocamlformat`. If another more generic tool comes along then we could always use that instead in Dune. Although it seems to me that `ocamlformat` has the right name for the job.

diml commented on src/format_rules.ml:
```diff
@@ -0,0 +1,81 @@
+  let open Build.O in
+  let action = Action.diff input output in
+  let alias_conf_name = "fmt" in
+  let stamp = ("format-alias", action) in
```
Instead of the action, we can just use `input` here, it'll be quicker to hash. We just need something that's unique.

diml commented on src/format_rules.ml:
```diff
@@ -0,0 +1,81 @@
+  in
+  let programs = Hashtbl.create 3 in
+  let resolve_program program_name =
+    Hashtbl.find_or_add
```
It feels like this caching should go directly in `Super_context.resolve_program`. In particular, we resolve `refmt` in `preprocessing.ml` as well.

emillon replied:
Done. On the test suite, the cache hit rate is 63%.

emillon commented on src/dune_project.ml:
```diff
@@ -364,33 +397,51 @@ let parse ~dir ~lang ~packages ~file =
+             let Extension.Extension e = extension in
+             let args =
+               let%map (arg, stanzas) = Dsexp.Of_sexp.set_many parsing_context e.stanzas in
+               let new_args_acc =
```
It's a bit tricky to do because parse_args can only pass a `Univ_map.t`. So, that works if we return a singleton and then add it or not (that's what I just pushed). Otherwise, we need to bring make `instance` polymorphic and add an existencial. That will require new types and annotations, I'm not sure it will be better. Is there an intermediate solution I'm missing?

diml replied:
Ah, that's right. Sorry, I missed that.

emillon replied:
Do you prefer the singleton version or the former one with `add`?

diml replied:
The former one looked better to me

emillon commented:
Regarding the benchmark, I ran two tests: first, the one you suggested. Null build with `-j 1`. Both with using `master` and this branch. But this is not super relevant because we don't iterate on the files when this feature is not enabled in the project. So I measured what happens when instead of ignoring the projects by default, we format them.

### :bar_chart: Results

case               | null build time (s)
------------------ | -------------------
master             | 1.35
opt-in per project |1.33
always format      | 1.48

Does that look acceptable? If that's not, how can I fix this? I understand that each rule has a directory, but I'm not sure how that translates in terms of when a rule is computed. Also, when there are multiple directories, say `a` and `b`, does that mean that we need to setup `a/.formatted` and `b/.formatted`, or rather `.formatted/a` and `.formatted/b`? Thanks!

diml commented:
Thanks for the benchmark. The difference seems enough to me to justify computing the rules only when needed. To do that, you need to use `Build_system.on_load_dir`. i.e. when generating the rules for `a`, you register a callback with `on_load_dir` for `a/.formatted` that will produce all the  formatting rules.

> Also, when there are multiple directories, say a and b, does that mean that we need to setup a/.formatted and b/.formatted, or rather .formatted/a and .formatted/b? 

You can use either, though the former looks more natural to me.

emillon commented:
I tried to do that, but I'm having some trouble. When I call `on_load_dir ~dir:"a"`,  I don't get any performance improvement (see last commit), but when I call `on_load_dir ~dir:"a/.formatted"`, nothing happens (the callback is not called, and the fmt alias does not get registered).

I assume that it's because nothing loads the directory. I tried to add `Super_context.add_alias_deps sctx alias (Path.Set.singleton output_dir)` but that does not fix the problem. Any idea how to fix this?

emillon committed:
Integration with auto formatters

This adds a "fmt" extension in dune-project files. When used, it will
setup a `@fmt` alias that will call `ocamlformat` on ocaml source code,
and `refmt` on reason source code. The tools are not configured by dune.

Closes #1201

Signed-off-by: Etienne Millon <me@emillon.org>

emillon commented:
Thanks for the pointer (didn't think to try with the actual contents instead of the directory). I just did that and now I have the following error (see latest commit):

> Build_system.get_collector called on closed directory

Any idea how to fix this?

diml commented:
Ah, it's because the diff actions are attached to `@a/fmt` while generating the rules for directory `a/.formatted`. In fact, `@a/fmt` needs to depend on `@a/.formatted/fmt` and the diff actions should be attached `@a/.formatted/fmt`.

emillon commented:
Thanks! That was it. This works and the performance is now comparable between the "opt in" and "always" cases. I think that this is ready.

emillon commented:
About config files: this is passing `--name` (so that the correct `.ocamlformat` files are used) and adding a dependency to `.ocamlformat` in the current directory if it exists. So, if a `.ocamlformat` from a parent directory is modified, no reformatting will be triggered.

This seems to be an acceptable to me in a first version. We can see how it's possible to improve this - possibly by extending `ocamlformat` so that it can return the set of actual files relevant to a directory.

diml commented:
> About config files: this is passing --name (so that the correct .ocamlformat files are used) and adding a dependency to .ocamlformat in the current directory if it exists. So, if a .ocamlformat from a parent directory is modified, no reformatting will be triggered.

Do users of ocamlformat put a single `.ocamlformat` file at the toplevel of their project or one in each directory? Intuitively, I would expect the former so that the style is set once and for all for the whole project.

emillon commented:
I think that a single one is the common practice, but it's possible to use this cascading feature to ignore a directory for example.

diml commented:
Ok. Technically all the files ocamlformat reads need to be added as dependency, otherwise `dune fmt` will give different results compared to what the user gets in the editor.

emillon commented:
We pass the original file (in the source tree) as --name, so the config
files are read from there. I think that it guarantees that it will do the
same as in the editor.

diml commented on src/format_rules.ml:
```diff
@@ -0,0 +1,101 @@
+             [ A (flag_of_kind kind)
+             ; Dep input
+             ; A "--name"
+             ; Path file
```
`Path.t` values in argument specification eventually get translated to: `Path.reach <path> ~from:<dir-where-the-command-is-executed>`. So for a file `src/foo.ml`, I believe this will get turned to `../../../src/foo.ml` given that the command is run in `_build/default/src`. I suggest to execute the command from `_build/default` (available via `(Super_context.context sctx).build_dir` and simply pass `input` here.

emillon replied:
Yes, that's what's passed (this appears in the test output). I'm not sure what we gain from changing the directory. This has to refer to both the build and the source directory, so we'll have to do some path manipulation anyway.

diml commented:
Ah ok, I get why we pass the file in the source tree via `--name` now. However, in case of error the filename reported in the error message will be odd. Why not add dependencies on all the `.ocamlformat` files up to the root of the workspace?

diml commented:
We can use `Build.paths_glob` or `Build.if_file_exists` to add a dependency only if the file exists.

emillon commented:
`Build.paths_glob` seems to be a nice solution, but if you don't mind can we do this in a separate PR? The current solution is already "correct" in the sense that we pick the correct files - just the incremental aspect can be improved. This PR has been opened for 3 weeks, I'd be grateful if we could merge the main part before improving it. Thanks!

diml commented:
Sure, sorry about the nitpicking.

emillon commented:
No problem! I filed two bugs so that I don't forget about this.

emillon commented:
Thanks for the thorough review!

emillon commented:
Yes! I've been running some tests about which configuration yields the smallest diff. I think our style corresponds to the "sparse" profile with some minor addition. I'll run the tests and post the results in a new issue.

hhugo commented:
I've mentioned that in a comment already (https://github.com/ocaml/dune/pull/1252#discussion_r216733768).
It would be nice to be able to specify exceptions on a per file basis. It's useful when a file do not parse (e.g file preprocessed with cppo) or when a file is automatically generated (e.g. by ocamllex, yacc, menhir, ...).

diml commented:
For generated files that are not promoted, we should probably disable auto-formatting automatically.

diml commented:
Actually, looking at the code that's already the case, so there should be no need to disable ocamlformat on files generated by ocamllex, yacc, ...

hhugo commented:
My use case is with lexers and parsers files that are promoted.
